inference-engine/src/vpu/graph_transformer/src/middleend/special_stage_processor.cpp

```diff
@@ -5,99 +5,249 @@
 #include "vpu/middleend/special_stage_processor.hpp"
 
 #include <vector>
-#include <set>
 #include <utility>
 
 namespace vpu {
 
-void SpecialStageProcessor::processSplit(
-        const Model& model,
-        const Stage& stage) {
-    IE_ASSERT(stage->type() == StageType::Split);
+namespace {
 
-    auto input = stage->input(0);
+struct NeedCopyDesc {
+    bool isCopyNeed = false;
+    vpu::Optional<bool> isCopyOptional;
+};
 
-    const auto& offsets = stage->attrs().get<std::vector<DimValues>>("offsets");
-    IE_ASSERT(offsets.size() == checked_cast<size_t>(stage->numOutputs()));
-
-    for (const auto& outEdge : stage->outputEdges()) {
-        IE_ASSERT(outEdge->portInd() >= 0);
-        IE_ASSERT(checked_cast<size_t>(outEdge->portInd()) < offsets.size());
+NeedCopyDesc isNeedCopyOfOutput(
+                const Stage& stage,
+                const StageOutput& outputEdge,
+                const Data& inputData) {
+    NeedCopyDesc needCopyDesc;
+    auto output = outputEdge->output();
+    if (output->usage() != DataUsage::Intermediate) {
+        needCopyDesc.isCopyNeed = true;
+    } else if (output->parentDataToDataEdge() != nullptr) {
+        needCopyDesc.isCopyNeed = true;
+    } else if (stage->type() != StageType::Reshape) {
+        // No need to check Reshape strides, because it must be compact
```

Comment:
After looking at original version for `Reshape` it seems to me it's easier to re-use only insertion helper there and don't call checkers (only if it's not a part of your fix)

Reply:
Ok, I think than you're right. Just my idea of refactoring is move all logic to one place, because I guess than it is easily to understand. When I have tried to refactor it I spent a lot of time to understand what's going on, but now you need take a look once to understand general logic and difference of all special stages.
I think it is good topic to discuss.

Reply:
Fixed accordingly with your point.